packages/react/scroll-area/src/ScrollArea.tsx

```diff
@@ -186,2234 +80,886 @@ type ScrollAreaPrimitive = Polymorphic.ForwardRefComponent<
 
 type ScrollAreaCornerImplOwnProps = Polymorphic.OwnProps<typeof Primitive>;
 type ScrollAreaCornerImplPrimitive = Polymorphic.ForwardRefComponent<
   Polymorphic.IntrinsicElement<typeof Primitive>,
   ScrollAreaCornerImplOwnProps
 >;
 
-const ScrollAreaCornerImpl = React.forwardRef(function ScrollAreaCornerImpl(props, forwardedRef) {
-  const { positionRef } = useScrollAreaRefs(CORNER_NAME);
-  const { dispatch } = useDispatchContext(CORNER_NAME);
-  const { dir } = useScrollAreaContext(CORNER_NAME);
-  const { domSizes } = useScrollAreaStateContext();
-  const isRTL = dir === 'rtl';
-
-  const style: any = {
-    // The resize handle is placed, by default, in the bottom right corner of the scroll area. In
-    // RTL writing mode it should be placed on the bottom left. Vertical reading modes have no
-    // impact on handle placement.
-    [SCROLL_AREA_CSS_PROPS.cornerLeft]: isRTL ? 0 : 'unset',
-    [SCROLL_AREA_CSS_PROPS.cornerRight]: isRTL ? 'unset' : 0,
-
-    [SCROLL_AREA_CSS_PROPS.cornerHeight]: domSizes.scrollbarX.height
-      ? domSizes.scrollbarX.height + 'px'
-      : domSizes.scrollbarY.width
-      ? domSizes.scrollbarY.width + 'px'
-      : '16px',
-    [SCROLL_AREA_CSS_PROPS.cornerWidth]: domSizes.scrollbarY.width
-      ? domSizes.scrollbarY.width + 'px'
-      : domSizes.scrollbarX.height
-      ? domSizes.scrollbarX.height + 'px'
-      : '16px',
-
-    position: 'absolute',
-  };
+const ScrollAreaCornerImpl = React.forwardRef((props, forwardedRef) => {
+  const context = useScrollAreaContext(CORNER_NAME);
+  const [width, setWidth] = React.useState<number>();
+  const [height, setHeight] = React.useState<number>();
+  const hasSize = Boolean(width && height);
 
-  /**
-   * The corner element is placed, by default, in the bottom right corner of the scroll area. In
-   * RTL writing mode it should be placed on the bottom left. Its cursor depends on the direction
-   * in which the user is able to resize the container.
-   *
-   * We rely on computed styles because the `resize` prop value can be `initial` or `inherit`
-   * which don't give us the direct information we need to derive the correct style for the
-   * cursor.
-   */
-  useLayoutEffect(() => {
-    if (positionRef.current) {
-      const computedStyles = getComputedStyle(positionRef.current);
-      dispatch({
-        type: ScrollAreaEvents.SetExplicitResize,
-        value: computedStyles.resize as ResizeBehavior,
-      });
-    }
-  }, [dispatch, positionRef]);
+  useResizeObserver(context.scrollbarX, () => {
+    const height = context.scrollbarX?.offsetHeight;
+    context.onCornerHeightChange(height || 0);
+    setHeight(height);
+  });
 
-  return (
+  useResizeObserver(context.scrollbarY, () => {
+    const width = context.scrollbarY?.offsetWidth;
+    context.onCornerWidthChange(width || 0);
+    setWidth(width);
+  });
+
+  return hasSize ? (
     <Primitive
       {...props}
       ref={forwardedRef}
       style={{
+        width,
+        height,
```

Comment:
I originally used the variables here too but we needed the `width` and `height` in state here to prevent the corner from rendering if there was no width/height (this was the change i made when you reported the bouncing corner circle bug).

Reply:
Yeah I thought that was why you did this, I guess that's probably fine, but it did make me think when reviewing the whole thing, especially seeing the state for corner size in 2 places, but obviously the top one is only used to set the css vars top-level and isn't being passed down. Maybe a comment on line 122 to explain we're using that state only to pass it through vars to avoid re-renders?